### DISCUSSION TITLE
Using a JavaScript `Image()` object with Next.js

### DISCUSSION THREAD

**Discussion by oconnoob** [vercel/next.js#44578] [Help] [answered]
Hi there,

I'm currently trying to use a ChartJS [Line chart](https://www.chartjs.org/docs/latest/charts/line.html) in Next.js via `react-chartjs-2`. Specifically, **I am trying to replace the data point circles on the chart with small images using the [`pointStyle`](https://www.chartjs.org/docs/3.7.1/samples/line/point-styling.html) property** by specifying a JavaScript `Image()` type as its value.

[Here is a codesandbox](https://codesandbox.io/s/radar-chart-react-shyv49?file=/src/App.js:0-1055) (and code below) of what I am trying to accomplish. This works fine in my Next.js app **on first render**, but if I refresh the page I run into an error - `ReferenceError: Image is not defined`.

I believe the error has something to do with Next's SSR, so I looked to the Next.js `<Image>` component's [loader](https://nextjs.org/docs/api-reference/next/image), but the only problem is that ChartJS cannot accept a Next.js `<Image>` type and requires a built-in JavaScript `Image()` object.

If anyone has any advice I would greatly appreciate it, thanks!

# Example Code

## `package.json`

```json
{
  "name": "frontend-next",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "lint": "next lint"
  },
  "dependencies": {
    "@emotion/react": "^11.10.5",
    "@emotion/styled": "^11.10.5",
    "@mui/material": "^5.10.15",
    "chart.js": "^4.1.1",
    "eslint": "8.28.0",
    "eslint-config-next": "13.0.5",
    "formik": "^2.2.9",
    "nanoid": "^4.0.0",
    "next": "13.0.5",
    "react": "18.2.0",
    "react-chartjs-2": "^5.1.0",
    "react-dom": "18.2.0",
    "yup": "^0.32.11"
  }
}

```

## Code

```javascript
import React from "react";
import {
  Chart as ChartJS,
  RadialLinearScale,
  PointElement,
  LineElement,
  Filler,
  Tooltip,
  Legend
} from "chart.js";
import { Radar } from "../node_modules/react-chartjs-2/dist";

function App() {
  ChartJS.register(
    RadialLinearScale,
    PointElement,
    LineElement,
    Filler,
    Tooltip,
    Legend
  );
  const myImage = new Image(25, 25);
  myImage.src = "https://i.stack.imgur.com/gXQrT.png";

  const data = {
    labels: ["Business 1", "Business 2", "Business 3"],
    datasets: [
      {
        label: "Number of Businesses",
        data: [1300, 400, 160],
        backgroundColor: "rgba(255, 99, 132, 0.2)",
        borderColor: "rgba(255, 99, 132, 1)",
        borderWidth: 1,
        pointStyle: myImage
      }
    ]
  };

  return (
    <div className="App">
      <Radar data={data} />

      <form>
        <input type="text" placeholder="Label" />
        <input type="text" placeholder="Dataset" />
        <button>Add Data</button>
      </form>
    </div>
  );
}

export default App;
```

**Comment by icyJoseph** [comment, marked as answer]
Right, so the Image API is not the same thing as the `next/image` React component.

Rather, new Image is more on the ballpark of `document.createElement('img')` - and it is unfortunately not available in Node.js.

According to Chart.js:

> If the value is an image or a canvas element, that image or canvas element is drawn on the canvas using drawImage

And in turn `drawImage`:

> An element to draw into the context. The specification permits any canvas image source, specifically, an HTMLImageElement, an SVGImageElement, an HTMLVideoElement, an HTMLCanvasElement, an ImageBitmap, an OffscreenCanvas, or a VideoFrame.

So that means that you cannot really pass a React Element, which you must remember is an Object describing your UI, a call to `React.createElement`, to keep it simple.

<img width="1141" alt="Screenshot 2023-01-05 at 09 40 09" src="https://user-images.githubusercontent.com/21013447/210737461-c38a292d-6a81-4a4b-8a5f-b7710f7a7dcf.png">

> Notice the above invokes, `React.createElement` not `document.createElement`

So what you need to do is pass a DOM object to `pointStyle`, and during SSR that's very much not possible, because there's no DOM. In fact, Chart.js must be using some kind of `canvas` compatible API for Node.js to allow it to work during SSR, or it might also crash. I must say I am not so experienced with that library to tell precisely.

My take on this would be to defer rendering to the client:

```js
import React, { useEffect, useRef, useState } from "react";
import {
  Chart as ChartJS,
  RadialLinearScale,
  PointElement,
  LineElement,
  Filler,
  Tooltip,
  Legend
} from "chart.js";
import { Radar } from "../node_modules/react-chartjs-2/dist";

// this really doesn't need to be within App I believe
ChartJS.register(
  RadialLinearScale,
  PointElement,
  LineElement,
  Filler,
  Tooltip,
  Legend
);

function App() {
  // const myImage = new Image(25, 25);
  // myImage.src = "https://i.stack.imgur.com/gXQrT.png";

  const ref = useRef();
  const [onClient, setOnClient] = useState();

  useEffect(() => {
    setOnClient(true);
  }, []);

  useEffect(() => {
    ref.current = new Image(25, 25);
    ref.current.src = "https://i.stack.imgur.com/gXQrT.png";
  }, []);

  const data = {
    labels: ["Business 1", "Business 2", "Business 3"],
    datasets: [
      {
        label: "Number of Businesses",
        data: [1300, 400, 160],
        backgroundColor: "rgba(255, 99, 132, 0.2)",
        borderColor: "rgba(255, 99, 132, 1)",
        borderWidth: 1,
        pointStyle: ref.current
      }
    ]
  };

  return (
    <div className="App">
      {onClient && <Radar data={data} />}

      <form>
        <input type="text" placeholder="Label" />
        <input type="text" placeholder="Dataset" />
        <button>Add Data</button>
      </form>
    </div>
  );
}

export default App;
```

Something like the above. I am not sure there's any replacement for the Web API `Image`, for Node.js in `npm`, but the above should be fine.

**Reply by oconnoob** [reply]
Thank you very much for this detailed answer! I understood that a `next/image` is not the same as `new Image()`, but I didn't realize the latter referred explicitly to the DOM and was not simply a JS type. Thank you for that clarification.

I've put my understanding below, with `(?)` next to things I think I could easily be wrong about. Would you mind checking this for posterity/my own edification?

### Problem
1. `new Image()` creates a DOM element in vanilla JS
2. `pointStyle` requires such a DOM object so that it can draw to its canvas
3. Since we are using SSR, we have no DOM to attach an object to

### Solution
1. Since our app function runs each time we re-render, we have no way to keep track of a specific DOM element given that the DOM changes with each render (?)
2. The `useRef()` hook can be used to address this. We create a `ref` and set its current value to `pointStyle`, which gives us a way to persist the object of reference between renders
3. When the `App()` component is first mounted to the DOM on the client side, we use two `useEffect` hooks. The first sets the `onClient` state to true, and the second sets the value of `ref.current` to the image.
4. In our return, we raplace `<Radar />` with `{onClient && <Radar />}`. This means that the `<Radar />` component will only be returned after mounting in the DOM on the client side, which prevents it from being returned on the server side, during which it would reference an undefined object - `ref.current` (?)

### Final questions
Finally, I have two remaining questions if you don't mind:
1. First, why does the original code work upon first render and only fails **after reloading**?
5. Second, in my implementation I'm actually returning the `<Radar />` component inside a `useMemo()` so that the graph only updates when the data passed into it updates. Does this additional detail change anything regarding your solution?

Thanks again for your insight!

**Reply by icyJoseph** [reply]
> First, why does the original code work upon first render and only fails after reloading?

Well I understand that although the link you shared if a CRA sandbox, and because of that I cannot guess why the first render is not crashing out on SSR already. I'd need to see more of your actual implementation.

> Second, in my implementation I'm actually returning the <Radar /> component inside a useMemo() so that the graph only updates when the data passed into it updates. Does this additional detail change anything regarding your solution?


Well, you could wrap Rader with `memo` instead.

```tsx
import {memo} from 'react';

const MemoRadar = memo(Radar);
```

And use the memoized version instead.
